### ISSUE TITLE
Dropdown menu accessibility issue from the beginning of time

### ISSUE THREAD

**Issue by marcoscv-work** [liferay/clay#5026]
We really need to move the dropdown options to the button of the page and make it visually positioned on the top of any other element, thats a Liferay need for some _corner cases_ (dropdown inside overflow hidden divs, sidebar menus, etc).

It caused issues related to accessibility and we have been fixing the focus during years, currently it is working fine on screen readers like VoiceOver but it still broken for any assistive technology based on the accessibility tree witch still broken because we don't show the dropdown menu options next to the trigger, a good example for this is JAWS, the most popular screen reader.

@victorg1991 has been fighting with it on this PTR: https://issues.liferay.com/browse/PTR-3245
We suggested to use virtual cursor but it could help for JAWS but not for other assistive technology based on the accessibility tree.

Here the accessibility tree showing the dropdown opened:
<img width="508" alt="image" src="https://user-images.githubusercontent.com/7963804/183052628-fe6e916f-21d2-4a08-9abc-d8b21958fe74.png">

**Proposal**:

Make the positioning on the page button configurable by API and true by default.
The "positioning on the page button"=false should position the dropdown next to the dropdown trigger, then we could change to false the cases one by one and fix this issue for most of the dropdowns

**Comment by matuzalemsteles** [comment]
Yeah this is a pretty old issue, we've had some talk about it over time, some in #2288, #2246 I believe there's even more talk about it in other issues, I also remember talking about making it configurable too and Chema had a few remarks about that too that wouldn't be ideal.

Regarding the problem I think this doesn't solve the problem with screen readers mainly JAWS and NVDA, I did a lot of testing with the fix in #4998, one of the things I noticed when using focus navigation is not related to move the focus of the button for the menu and yes the use of `role` with actions for the elements so that it can do the navigation correctly.

This PR #4998 adds these actions by default but they are configurable, I think the main problem is in how we organize the markup in the menu with the actions mainly for the group creation, the creation for example for Radios inside the Menu works well because we added `aria-label` to the element with `role="radiogroup"` with the same value of the group title , but this is not done for the other elements because the markup does not handle it correctly and causes JAWS or NVDA not to read the "Page Set" in the ticket example.

For example, here's what the group markup looks like in the Item Menu:

```html
<ul class="list-unstyled" role="menu">
  <li class="dropdown-subheader" role="presentation">Group #1</li>
  <li role="none">
    <a class="dropdown-item" href="#one" role="menuitem">one</a>
  </li>
  <li role="none">
    <a class="dropdown-item" href="#two" role="menuitem">two</a>
  </li>
  <li role="none">
    <a class="dropdown-item" href="#three" role="menuitem">three</a>
  </li>
</ul>
```

The items have no connection with the group title, to do that we would have to change the group markup to something like:

```html
<ul class="list-unstyled" role="menu">
  <li role="presentation">
    <span id="clay-id-1" aria-hidden="true">Group #1</span>
    <ul role="group" aria-labelledby="clay-id-1" class="list-unstyled">
      <li role="none">
        <a class="dropdown-item" href="#one" role="menuitem">one</a>
      </li>
      <li role="none">
        <a class="dropdown-item" href="#two" role="menuitem">two</a>
      </li>
      <li role="none">
        <a class="dropdown-item" href="#three" role="menuitem">three</a>
      </li>
    </ul>
  </li>
</ul>

// In order not to break the markup we can also have the same effect

<ul class="list-unstyled" role="menu">
  <li id="clay-id-1" class="dropdown-subheader" role="presentation">Group #1</li>
  <li role="presentation">
    <ul role="group" aria-labelledby="clay-id-1">
      <li role="none">
        <a class="dropdown-item" href="#one" role="menuitem">one</a>
      </li>
      <li role="none">
        <a class="dropdown-item" href="#two" role="menuitem">two</a>
      </li>
      <li role="none">
        <a class="dropdown-item" href="#three" role="menuitem">three</a>
      </li>
    </ul>
  </li>
</ul>
```

I think we always focused on solving just the focus for the visual and then we started to worry about accessibility but just testing with screen readers like VoiceOver but as JAWS and NVDA follow different behavior we started to make improvements to that, I think the next step to improve this is to solve the markup for groups.

**Comment by marcoscv-work** [comment]
Role/aria examples is not working for my on JAWS, maybe we are not using the tools in the same way than people with stoppers or disabilities on those technologies, they usually uses tree navigation, not focusing navigation, probably best solution is always follow the accessibility tree to ensure the functionality with all technologies, what is the issue exactly.

One the other hand, Even if we could fix this by ARIA one of the main rules in ARIA is **not to use ARIA if you can do it in a standard way**, in our case: ordering the HTML items in the natural order, not breaking the accessibility tree order and trying to fixit it  if someone reports something related

**Comment by matuzalemsteles** [comment]
Well I'm not sure but I believe the settings can be different for each type of person especially if it's configurable by the screen reader, so I wouldn't say there is a "correct configuration" way.

But anyway I'm still not sure that setting the button with the menu following the hierarchy is the root problem mainly for the ticket case, still we have to use ARIA correctly so that screen readers can read it, this is independent of order from the DOM or if the semantics is correct the idea is to try to add context, so we are using `role="menu"`, `role="menuitem"`, `role="group"`... that's the way it is just provide context to the elements even though we use the semantics correctly so I think that Victor's case is related to the incorrect use of aria to say the group of items within DropDown and this is still not related to the hierarchy of the button and menu.

Anyway I did some more tests on the main screen readers, VoiceOver, JAWS and NVDA just testing if we have hierarchy issues when you move the focus from the button to the menu, this works very well you may notice that the screen reader does not read the title of the group exactly because we are not saying that the items are grouped and it is also being ignored.

The tests follow the default configuration, the navigation of all seems to follow the top down order, so having the menu not being rendered in the same hierarchy is not a clear problem by the tests because we fixed the focus management, so for the screen reader it's like if it was in the same hierarchy.

Tests with JAWS and NVDA, have sound, so it can help to better identify the behavior, I also added to show the log.

### Mac VoiceOver

https://user-images.githubusercontent.com/13750819/183747633-5673cdb3-c579-4005-b995-5e02e325cbf2.mov

### Windows JAWS

https://user-images.githubusercontent.com/13750819/183755528-64260c12-4dce-4657-b2a0-f31b4e37363a.mp4

### Windows NVDA

https://user-images.githubusercontent.com/13750819/183751559-c470b6f9-035b-4b75-a8be-8ef232edaaff.mp4
